internal/command/add.go

```diff
@@ -0,0 +1,330 @@
+	}
+
+	// Get the schemas from the context
+	if _, exists := schemas.Providers[absProviderConfig.Provider]; !exists {
+		diags = diags.Append(tfdiags.Sourceless(
+			tfdiags.Error,
+			"Missing schema for provider",
+			fmt.Sprintf("No schema found for provider %s. Please verify that this provider exists in the configuration.", absProviderConfig.Provider.String()),
+		))
+		c.View.Diagnostics(diags)
+		return 1
+	}
+
+	// Get the schema for the resource
+	schema, schemaVersion := schemas.ResourceTypeConfig(absProviderConfig.Provider, rs.Mode, rs.Type)
+	if schema == nil {
+		diags = diags.Append(tfdiags.Sourceless(
+			tfdiags.Error,
+			"Missing resource schema from provider",
+			fmt.Sprintf("No resource schema found for %s.", rs.Type),
+		))
+		c.View.Diagnostics(diags)
+		return 1
+	}
+
+	stateVal := cty.NilVal
+	// Now that we have the schema, we can decode the previously-acquired resource state
+	if args.FromResourceAddr != nil {
+		ri := resource.Instance(args.FromResourceAddr.Resource.Key)
+		if ri.Current == nil {
+			diags = diags.Append(tfdiags.Sourceless(
+				tfdiags.Error,
+				"No state for resource",
+				fmt.Sprintf("There is no state found for the resource %s, so add cannot populate values.", rs.String()),
+			))
+			c.View.Diagnostics(diags)
+			return 1
+		}
+
+		rio, err := ri.Current.Decode(schema.ImpliedType())
+		if err != nil {
+			diags = diags.Append(tfdiags.Sourceless(
+				tfdiags.Error,
+				"Error decoding state",
+				fmt.Sprintf("Error decoding state for resource %s: %s", rs.String(), err.Error()),
+			))
+			c.View.Diagnostics(diags)
+			return 1
+		}
+
+		if ri.Current.SchemaVersion != schemaVersion {
+			diags = diags.Append(tfdiags.Sourceless(
+				tfdiags.Error,
+				"Schema version mismatch",
+				fmt.Sprintf("schema version %d for %s in state does not match version %d from the provider", ri.Current.SchemaVersion, rs.String(), schemaVersion),
+			))
+			c.View.Diagnostics(diags)
+			return 1
+		}
+
+		stateVal = rio.Value
+	}
+
+	diags = diags.Append(view.Resource(args.Addr, schema, localProviderConfig, stateVal))
+	if diags.HasErrors() {
+		c.View.Diagnostics(diags)
+		return 1
+	}
+
+	return 0
+}
+
+func (c *AddCommand) Help() string {
+	helpText := `
+Usage: terraform [global options] add [options] ADDRESS
+
+  Generates a blank resource template. With no additional flags,
+  the template will be displayed in the terminal. 
+
+Options:
+
+-from-state=ADDRESS		Fill the template with values from an existing resource.
+                        The resource must be the same type as the target address
+                        and exist in state.
+
+-out=string 			Write the template to a file. If the file already
+                        exists, the template will be appended to the file.
+
+-optional=false			Include optional attributes. Defaults to false.
+
+-provider=provider		Override the configured provider for the resource. Conflicts
+                        with -from-state
```

Comment:
It looks like there's some mixed tabs and spaces here that are making this not line up in situations where the tab stop size != 4. I think we typically uses spaces exclusively when we're formatting these option help blocks, to avoid dealing with that problem.